modules/gapi/src/backends/fluid/gfluidcore.cpp

```diff
@@ -843,6 +971,297 @@ static void run_arithm_s(DST out[], const SRC in[], int width, int chan,
+    float init[5] = { *(scalar + 1), *(scalar + 2), *scalar, *(scalar + 1) , *(scalar + 2) };
+    v_float32 s1 = v_set_scalar(scalar, chan),
+              s2 = v_set_scalar(init, chan),
+              s3 = v_set_scalar(init + 1, chan);
+    int length = width * 3;
+
+    if (std::is_same<T, uchar>::value)
+    {
+        constexpr int nlanes = static_cast<int>(v_uint8::nlanes);
+
+        if (length < 3 * nlanes)
+            return 0;
+
+        int x = 0;
+        v_float32 vectors[12];
+        for (;;)
+        {
+            for (; x <= length - 3 * nlanes; x += 3 * nlanes)
+            {
+                for (int i = 0; i < 12; ++i)
+                {
+                    vectors[i] = v_setr(reinterpret_cast<const uchar*>(in + x + i * nlanes / 4));
+                }
+
+                vx_store(reinterpret_cast<uchar*>(&out[x]), v_pack_u(v_pack(v_round(v_absdiff(vectors[0], s1)),
+                                                                            v_round(v_absdiff(vectors[1], s2))),
+                                                                     v_pack(v_round(v_absdiff(vectors[2], s3)),
+                                                                            v_round(v_absdiff(vectors[3], s1)))));
+
+                vx_store(reinterpret_cast<uchar*>(&out[x + nlanes]), v_pack_u(v_pack(v_round(v_absdiff(vectors[4], s2)),
+                                                                                     v_round(v_absdiff(vectors[5], s3))),
+                                                                              v_pack(v_round(v_absdiff(vectors[6], s1)),
+                                                                                     v_round(v_absdiff(vectors[7], s2)))));
+
+                vx_store(reinterpret_cast<uchar*>(&out[x + 2 * nlanes]), v_pack_u(v_pack(v_round(v_absdiff(vectors[8], s3)),
+                                                                                         v_round(v_absdiff(vectors[9], s1))),
+                                                                                  v_pack(v_round(v_absdiff(vectors[10], s2)),
+                                                                                         v_round(v_absdiff(vectors[11], s3)))));
+            }
+
+            if (x < length && (in != out))
+            {
+                x = length - 3 * nlanes;
+                continue;  // process one more time (unaligned tail)
+            }
+            break;
+        }
+        return x;
+    }
+    else if (std::is_same<T, ushort>::value)
+    {
+        return absdiffc_simd_shorts_c3(reinterpret_cast<const ushort*>(in),
+                                       reinterpret_cast<ushort*>(out), s1, s2, s3, length);
+    }
+    else if (std::is_same<T, short>::value)
+    {
+        return absdiffc_simd_shorts_c3(reinterpret_cast<const short*>(in),
+                                       reinterpret_cast<short*>(out), s1, s2, s3, length);
+    }
+    return 0;
+}
```

Comment:
All these `reinterpret_cast`s don't look good, especially assuming they are not really casting anything to anywhere.. I've tried to remove them, and they turned out to be "necessary" in this code structure.
To avoid this problem and remove the casts, I suggest to separate `absdiffc_simd_c3` function template specialization for `uchar` - and then you won't need to cast anything there.
Also this step will allow to delete the if-elses for the `T` type inside this function

This compiles, if you like to use it:
```
template<typename T>
CV_ALWAYS_INLINE int absdiffc_simd_c3(const T in[], const float scalar[], T out[], int width)
{
    bool isShort = std::is_same<T, ushort>::value || std::is_same<T, short>::value;
    GAPI_Assert(isShort == true);

    constexpr int chan = 3;
    float init[5] = { *(scalar + 1), *(scalar + 2), *scalar, *(scalar + 1) , *(scalar + 2) };
    v_float32 s1 = v_set_scalar(scalar, chan),
              s2 = v_set_scalar(init, chan),
              s3 = v_set_scalar(init + 1, chan);
    int length = width * 3;

    return absdiffc_simd_shorts_c3(in, out, s1, s2, s3, length);
}

template<>
CV_ALWAYS_INLINE int absdiffc_simd_c3<uchar>(const uchar in[], const float scalar[], uchar out[], int width)
{
    constexpr int chan = 3;
    float init[5] = { *(scalar + 1), *(scalar + 2), *scalar, *(scalar + 1) , *(scalar + 2) };
    v_float32 s1 = v_set_scalar(scalar, chan),
              s2 = v_set_scalar(init, chan),
              s3 = v_set_scalar(init + 1, chan);
    int length = width * 3;

    constexpr int nlanes = static_cast<int>(v_uint8::nlanes);

    if (length < 3 * nlanes)
        return 0;

    int x = 0;
    v_float32 vectors[12];
    for (;;)
    {
        for (; x <= length - 3 * nlanes; x += 3 * nlanes)
        {
            for (int i = 0; i < 12; ++i)
            {
                vectors[i] = v_setr(in + x + i * nlanes / 4);
            }

            vx_store(&out[x], v_pack_u(v_pack(v_round(v_absdiff(vectors[0], s1)),
                                                                        v_round(v_absdiff(vectors[1], s2))),
                                                                    v_pack(v_round(v_absdiff(vectors[2], s3)),
                                                                        v_round(v_absdiff(vectors[3], s1)))));

            vx_store(&out[x + nlanes], v_pack_u(v_pack(v_round(v_absdiff(vectors[4], s2)),
                                                                                    v_round(v_absdiff(vectors[5], s3))),
                                                                            v_pack(v_round(v_absdiff(vectors[6], s1)),
                                                                                    v_round(v_absdiff(vectors[7], s2)))));

            vx_store(&out[x + 2 * nlanes], v_pack_u(v_pack(v_round(v_absdiff(vectors[8], s3)),
                                                                                        v_round(v_absdiff(vectors[9], s1))),
                                                                                v_pack(v_round(v_absdiff(vectors[10], s2)),
                                                                                        v_round(v_absdiff(vectors[11], s3)))));
        }

        if (x < length && (in != out))
        {
            x = length - 3 * nlanes;
            continue;  // process one more time (unaligned tail)
        }
        break;
    }
    return x;
}
```

Reply:
This is similar to what Ruslan suggested for a different function above, so I hope it will be helpful :)

Reply:
Done

Reply:
Great!